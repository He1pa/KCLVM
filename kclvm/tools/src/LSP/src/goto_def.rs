use kclvm_ast::ast::{AssignStmt, Program, Stmt, UnificationStmt};
use kclvm_ast::pos::ContainsPos;
use kclvm_error::Position as KCLPos;
use kclvm_sema::resolver::scope::ProgramScope;
use lsp_types::{GotoDefinitionResponse, Url};
use lsp_types::{Location, Range};

use crate::to_lsp::lsp_pos;

pub(crate) fn goto_definition(
    program: Program,
    kcl_pos: KCLPos,
    prog_scope: ProgramScope,
) -> Option<lsp_types::GotoDefinitionResponse> {
    match program.pos_to_stmt(&kcl_pos) {
        Some(node) => match node.node {
            Stmt::Unification(stmt) => goto_def_for_unification(stmt, kcl_pos, prog_scope),
            Stmt::Assign(stmt) => goto_def_for_assign(stmt, kcl_pos, prog_scope),
            _ => None,
        },
        None => None,
    }
}

pub(crate) fn goto_def_for_unification(
    stmt: UnificationStmt,
    kcl_pos: KCLPos,
    prog_scope: ProgramScope,
) -> Option<GotoDefinitionResponse> {
    let schema_expr = stmt.value.node;
    if schema_expr.name.contains_pos(&kcl_pos)

    {
        let id = schema_expr.name.node.names.last().unwrap();
        let positions = find_name_in_program_scope(id, prog_scope);
        positions_to_goto_def_resp(&positions)
    } else {
        None
    }
}

pub(crate) fn goto_def_for_assign(
    stmt: AssignStmt,
    kcl_pos: KCLPos,
    prog_scope: ProgramScope,
) -> Option<GotoDefinitionResponse> {
    let id = {
        if let Some(ty) = stmt.type_annotation {
            if ty.contains_pos(&kcl_pos) {
                Some(ty.node)
            } else {
                None
            }
        } else if stmt.value.contains_pos(&kcl_pos){
            match stmt.value.node {
                kclvm_ast::ast::Expr::Identifier(id) => Some(id.names.last().unwrap().clone()),
                kclvm_ast::ast::Expr::Schema(schema_expr) => {
                    if schema_expr.name.contains_pos(&kcl_pos) {
                        Some(schema_expr.name.node.names.last().unwrap().clone())
                    } else {
                        None
                    }
                },
                _ => None

            }
        } else {
            None
        }
    };
    match id {
        Some(id) => {
            let positions = find_name_in_program_scope(&id, prog_scope);
            positions_to_goto_def_resp(&positions)
        }
        None => None,
    }
}

pub(crate) fn find_name_in_program_scope(
    name: &String,
    prog_scope: ProgramScope,
) -> Vec<(KCLPos, KCLPos)> {
    let mut positions = vec![];
    let mut scopes = vec![];
    for s in prog_scope.scope_map.values() {
        scopes.push(s.borrow().clone());
    }

    while !scopes.is_empty() {
        let s = scopes.pop().unwrap();
        match s.lookup(&name) {
            Some(obj) => {
                let obj = obj.borrow().clone();
                positions.push((obj.start, obj.end));
            }
            None => {
                for c in s.children {
                    scopes.push(c.borrow().clone());
                }
            }
        }
    }
    positions
}

pub(crate) fn positions_to_goto_def_resp(
    positions: &Vec<(KCLPos, KCLPos)>,
) -> Option<GotoDefinitionResponse> {
    match positions.len() {
        0 => None,
        1 => {
            let (start, end) = positions[0].clone();
            Some(lsp_types::GotoDefinitionResponse::Scalar(Location {
                uri: Url::from_file_path(start.filename.clone()).unwrap(),
                range: Range {
                    start: lsp_pos(&start),
                    end: lsp_pos(&end),
                },
            }))
        }
        _ => {
            let mut res = vec![];
            for (start, end) in positions {
                res.push(Location {
                    uri: Url::from_file_path(start.filename.clone()).unwrap(),
                    range: Range {
                        start: lsp_pos(&start),
                        end: lsp_pos(&end),
                    },
                })
            }
            Some(lsp_types::GotoDefinitionResponse::Array(res))
        }
    }
}
