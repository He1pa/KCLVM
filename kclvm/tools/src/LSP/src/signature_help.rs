use crate::goto_def::find_def;
use kclvm_error::Position as KCLPos;
use kclvm_sema::core::global_state::GlobalState;
use kclvm_sema::core::scope::Scope;
use lsp_types::ParameterInformation;
use lsp_types::SignatureHelp;
use lsp_types::SignatureInformation;

pub fn signature_help(
    pos: &KCLPos,
    gs: &GlobalState,
    trigger_character: Option<String>,
) -> Option<SignatureHelp> {
    if trigger_character.is_none() {
        return None;
    }
    match trigger_character.unwrap().as_str() {
        // func<cursor>
        "(" => {
            if let Some(def) = find_def(pos, gs, false) {
                match def.get_kind() {
                    kclvm_sema::core::symbol::SymbolKind::Value => {
                        if let Some(symbol) = gs.get_symbols().get_symbol(def) {
                            if let Some(ty) = &symbol.get_sema_info().ty {
                                match &ty.kind {
                                    kclvm_sema::ty::TypeKind::Function(func_ty) => {
                                        let label = func_ty.func_signature_str(&symbol.get_name());
                                        let documentation = match &symbol.get_sema_info().doc {
                                            Some(s) => {
                                                Some(lsp_types::Documentation::String(s.clone()))
                                            }
                                            None => None,
                                        };
                                        let parameters = Some(
                                            func_ty
                                                .params
                                                .iter()
                                                .map(|param| ParameterInformation {
                                                    label: lsp_types::ParameterLabel::Simple(
                                                        format!(
                                                            "{}: {}",
                                                            param.name,
                                                            param.ty.ty_str()
                                                        ),
                                                    ),
                                                    documentation: None,
                                                })
                                                .collect::<Vec<ParameterInformation>>(),
                                        );

                                        return Some(SignatureHelp {
                                            signatures: vec![SignatureInformation {
                                                label,
                                                documentation,
                                                parameters,
                                                active_parameter: Some(0),
                                            }],
                                            active_signature: None,
                                            active_parameter: None,
                                        });
                                    }
                                    _ => {}
                                }
                            }
                        }
                    }
                    _ => {}
                }
            }

            None
        }
        // func(arg1,<cursor>)
        "," => {
            if let Some(scope) = gs.look_up_scope(pos) {
                match scope.get_kind() {
                    kclvm_sema::core::scope::ScopeKind::Local => {
                        if let Some(local_scope) = gs.get_scopes().try_get_local_scope(&scope) {
                            match local_scope.get_kind() {
                                kclvm_sema::core::scope::LocalSymbolScopeKind::Callable => {
                                    if let Some(func_symbol) = local_scope.get_owner() {
                                        if let Some(symbol) =
                                            gs.get_symbols().get_symbol(func_symbol)
                                        {
                                            if let Some(ty) = &symbol.get_sema_info().ty {
                                                match &ty.kind {
                                                    kclvm_sema::ty::TypeKind::Function(func_ty) => {
                                                        let label = func_ty
                                                            .func_signature_str(&symbol.get_name());
                                                        let documentation = match &symbol
                                                            .get_sema_info()
                                                            .doc
                                                        {
                                                            Some(s) => Some(
                                                                lsp_types::Documentation::String(
                                                                    s.clone(),
                                                                ),
                                                            ),
                                                            None => None,
                                                        };
                                                        let parameters = Some(
                                                            func_ty
                                                                .params
                                                                .iter()
                                                                .map(|param| ParameterInformation {
                                                                    label:
                                                                        lsp_types::ParameterLabel::Simple(
                                                                            format!(
                                                                                "{}: {}",
                                                                                param.name,
                                                                                param.ty.ty_str()
                                                                            ),
                                                                        ),
                                                                    documentation: None,
                                                                })
                                                                .collect::<Vec<ParameterInformation>>(),
                                                        );

                                                        return Some(SignatureHelp {
                                                            signatures: vec![
                                                                SignatureInformation {
                                                                    label,
                                                                    documentation,
                                                                    parameters,
                                                                    active_parameter: None,
                                                                },
                                                            ],
                                                            active_signature: None,
                                                            active_parameter: None,
                                                        });
                                                    }
                                                    _ => {}
                                                }
                                            }
                                        }
                                    }
                                }
                                _ => {}
                            }
                        }
                    }
                    kclvm_sema::core::scope::ScopeKind::Root => {}
                }
            }
            None
        }
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::signature_help;

    use crate::tests::compile_test_file;
    use kclvm_error::Position as KCLPos;

    #[test]
    fn aaaa() {
        let (file, _program, _, gs) =
            compile_test_file("src/test_data/signature_help/signature_help.k");

        let pos = KCLPos {
            filename: file.clone(),
            line: 5,
            column: Some(9),
        };
        let got = signature_help(&pos, &gs, Some("(".to_string())).unwrap();
        println!("{:?}", got);

        let pos = KCLPos {
            filename: file.clone(),
            line: 6,
            column: Some(10),
        };
        let got = signature_help(&pos, &gs, Some(",".to_string())).unwrap();
        println!("{:?}", got);
    }
}
