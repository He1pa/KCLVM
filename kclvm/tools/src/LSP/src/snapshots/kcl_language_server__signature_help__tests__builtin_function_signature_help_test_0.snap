---
source: tools/src/LSP/src/signature_help.rs
expression: "format!(\"{:#?}\", res)"
---
SignatureHelp {
    signatures: [
        SignatureInformation {
            label: "function pow(x: int | float, y: int | float, z: int | float) -> int | float",
            documentation: Some(
                String(
                    "Equivalent to `x ** y` (with two arguments) or `x ** y % z` (with three arguments)\n\nSome types, such as ints, are able to use a more efficient algorithm when\ninvoked using the three argument form.",
                ),
            ),
            parameters: Some(
                [
                    ParameterInformation {
                        label: Simple(
                            "x: int | float",
                        ),
                        documentation: None,
                    },
                    ParameterInformation {
                        label: Simple(
                            "y: int | float",
                        ),
                        documentation: None,
                    },
                    ParameterInformation {
                        label: Simple(
                            "z: int | float",
                        ),
                        documentation: None,
                    },
                ],
            ),
            active_parameter: Some(
                0,
            ),
        },
    ],
    active_signature: None,
    active_parameter: None,
}
